### ISSUE TITLE
[BUG] Problems installing and running getting started notebooks

### ISSUE THREAD

**Issue by masoncusack** [NVIDIA-Merlin/models#962]
### Bug description
Dependency conflict on pip install

### Steps/Code to reproduce bug

1. Create and activate rapids conda environment with `conda create -y -n rapids -c rapidsai -c nvidia -c conda-forge rapids=22.06 python=3.8 cudatoolkit=11.2`
2. Pip install as instructed in README `pip install merlin-models`

### Expected behavior
```bash
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
dask-cudf 22.6.1 requires cupy-cuda115, which is not installed.
cudf 22.6.1 requires cupy-cuda115, which is not installed.
cudf-kafka 22.6.1 requires cython, which is not installed.
```

Running the first cell of the getting started notebook, Tensorflow is then missing

```
ModuleNotFoundError: No module named 'tensorflow'
```

I try to pip install this with `pip install tensorflow>=2.8`, and the this fails with the same error:

```
(rapids-merlin-models) azureuser@mason-v100-new:~/cloudfiles/code/Users/mason.cusack$ pip install tensorflow>=2.8
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf 22.6.1 requires cupy-cuda115, which is not installed.
cudf-kafka 22.6.1 requires cython, which is not installed.
cudf 22.6.1 requires protobuf<3.21.0a0,>=3.20.1, but you have protobuf 3.19.6 which is incompatible.
```

### Environment details
- Merlin version: merlin-core==0.10.0, merlin-models==0.10.0
- Platform: Ubuntu 20.04
- Python version: 3.8
- PyTorch version (GPU?):
- Tensorflow version (GPU?): 2.11.0 (installed as in requirements >=2.8)

**Comment by masoncusack** [comment]
I then ran

```bash
pip install cupy-cuda115
pip install cython
```

And after this `pip install tensorflow>=2.8` runs without error.

The first cell of the getting started notebook, when run, then warns:

```
  CuPy may not function correctly because multiple CuPy packages are installed
  in your environment:

    cupy, cupy-cuda115
```

I followed the advice on the [webpage](https://docs.cupy.dev/en/stable/install.html) the warning links to, and because I have the cuda toolkit 11.2 installed, I uninstall other cupy installations and `pip install cupy-cuda11x`.

After this the first cell of the getting started notebook runs but the second fails with

```
ImportError: ('nvtabular is required for this feature.', 'Please install it with `pip install nvtabular`.')
```
So I pip install and the second cell runs, with the warning

```
Initializing an NVTabular Dataset in CPU mode.This is an experimental feature with extremely limited support!
```

But then fails with

```
NameError                                 Traceback (most recent call last)
Cell In[3], line 2
      1 input_path = os.environ.get("INPUT_DATA_DIR", os.path.expanduser("~/merlin-models-data/movielens/"))
----> 2 train, valid = get_movielens(variant="ml-1m", path=input_path)

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/datasets/entertainment/movielens/dataset.py:85, in get_movielens(path, variant, overwrite, transformed_name, nvt_workflow, **kwargs)
     83 nvt_path_exists = train_path.exists() and valid_path.exists()
     84 if not nvt_path_exists or overwrite:
---> 85     transform_movielens(
     86         raw_path, nvt_path, nvt_workflow=nvt_workflow, variant=variant, **kwargs
     87     )
     89 train = merlin.io.Dataset(str(train_path), engine="parquet")
     90 valid = merlin.io.Dataset(str(valid_path), engine="parquet")

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/datasets/entertainment/movielens/dataset.py:143, in transform_movielens(raw_data_path, output_path, nvt_workflow, variant, **kwargs)
    141     _nvt_workflow = default_ml25m_transformation(**locals())
    142 elif variant == "ml-1m":
--> 143     _nvt_workflow = default_ml1m_transformation(**locals())
    144 elif variant == "ml-100k":
    145     _nvt_workflow = default_ml100k_transformation(**locals())

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/datasets/entertainment/movielens/dataset.py:266, in default_ml1m_transformation(raw_data_path, **kwargs)
    258 joined = (
    259     ["userId", "movieId"]
    260     >> ops.JoinExternal(movies, on=["movieId"])
    261     >> ops.JoinExternal(users, on=["userId"])
    262 )
    264 cat = lambda: nvt.ops.Categorify(dtype="int32")  # noqa
--> 266 cat_features = joined >> cat()
    267 label = nvt.ColumnSelector(["rating"])
    269 # Columns to apply to

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/datasets/entertainment/movielens/dataset.py:264, in default_ml1m_transformation.<locals>.<lambda>()
    257 users = df_lib.read_parquet(os.path.join(raw_data_path, "users_converted.parquet"))
    258 joined = (
    259     ["userId", "movieId"]
    260     >> ops.JoinExternal(movies, on=["movieId"])
    261     >> ops.JoinExternal(users, on=["userId"])
    262 )
--> 264 cat = lambda: nvt.ops.Categorify(dtype="int32")  # noqa
    266 cat_features = joined >> cat()
    267 label = nvt.ColumnSelector(["rating"])

NameError: name 'nvt' is not defined
```

**Comment by masoncusack** [comment]
Hi @rnyak,

I'm still having these issues today. Going to see if there's any overlap from the issues resolved with Transformers4Rec that might help and report back.

I am a bit concerned that the nvt NameError might be a genuine problem.

Let me know if you have any immediate suggestions.

**Comment by masoncusack** [comment]
Import nvtabular works, I just get the warnings

  CuPy may not function correctly because multiple CuPy packages are installed
  in your environment:

    cupy, cupy-cuda11x

  Follow these steps to resolve this issue:

    1. For all packages listed above, run the following command to remove all
       existing CuPy installations:

         $ pip uninstall <package_name>

      If you previously installed CuPy via conda, also run the following:

         $ conda uninstall cupy

    2. Install the appropriate CuPy package.
       Refer to the Installation Guide for detailed instructions.

         https://docs.cupy.dev/en/stable/install.html

**Comment by masoncusack** [comment]
Import nvtabular works after pip install, but I'm still getting the NameError above.

Originally I had installed Tensorflow>=2.8 as in the requirements and 2.11.0 was installed. This seems to import fine and detects the GPU (do let me know if this was actually wrong).

I followed your instruction:
- pip install cudf-cu11 dask-cudf-cu11 --extra-index-url=https://pypi.ngc.nvidia.com
- pip install nvtabular
- pip install merlin-dataloader
- pip install merlin-core>=0.2.0
- pip install . --no-deps (with latest commit from source)
- pip install tensorflow==2.10 (originally I was doing pip install tensorflow>=2.8.0 as in requirements)

After this when running model.fit I get the error

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 model.fit(train, batch_size=1024)

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/models/tf/models/base.py:1127, in BaseModel.fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing, train_metrics_steps, pre, **kwargs)
   1101 def fit(
   1102     self,
   1103     x=None,
   (...)
   1124     **kwargs,
   1125 ):
   1126     x = _maybe_convert_merlin_dataset(x, batch_size, **kwargs)
-> 1127     self._maybe_set_schema(x)
   1129     validation_data = _maybe_convert_merlin_dataset(
   1130         validation_data, batch_size, shuffle=shuffle, **kwargs
   1131     )
   1133     callbacks = self._add_metrics_callback(callbacks, train_metrics_steps)

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/models/tf/models/base.py:1086, in BaseModel._maybe_set_schema(self, maybe_loader)
   1083 target_tags = [Tags.TARGET, Tags.BINARY_CLASSIFICATION, Tags.REGRESSION]
   1084 if self.input_schema:
   1085     loader_output_features = set(
-> 1086         loader.output_schema.excluding_by_tag(target_tags).column_names
   1087     )
   1088     model_input_features = set(self.input_schema.column_names)
   1089     schemas_match = loader_output_features == model_input_features

File /anaconda/envs/rapids-merlin-models/lib/python3.8/site-packages/merlin/models/tf/loader.py:335, in Loader.output_schema(self)
    333 @property
    334 def output_schema(self) -> Schema:
--> 335     output_schema = super().output_schema
    336     for map_fn in self._map_fns:
    337         if hasattr(map_fn, "compute_output_schema"):

AttributeError: 'super' object has no attribute 'output_schema'
```

I'll try running through the process again to make sure I didn't get anything wrong.

**Comment by masoncusack** [comment]
Followed the above again exactly in a new conda environment and I get the same AttributeError.

**Comment by rnyak** [comment]
@masoncusack what are the versions of the dataloader, core, nvtabular? can you please check and write here?

**Comment by masoncusack** [comment]
- nvtabular==1.5.0
- merlin-dataloader==0.0.4
- merlin-core==0.10.0

**Comment by jperez999** [comment]
Hello @masoncusack can you try pip installing your data loader with the following:
`pip install --upgrade git+https://github.com/NVIDIA-Merlin/dataloader.git --no-deps`
Let me know if that fixes your error.

**Comment by masoncusack** [comment]
Hi @jperez999, I've done that and the first notebook now seems to run all the way through.

I do get logging errors and the following warnings running `get_movielens`

```
Message: 'A GPU was detected but rapids is not installed.'
Arguments: ('NVTabular will not be able to use GPU.', 'Look at the documentation for more information at rapids.ai')
/anaconda/envs/rapids-dataloader-from-source/lib/python3.8/site-packages/merlin/io/dataset.py:253: UserWarning: Initializing an NVTabular Dataset in CPU mode.This is an experimental feature with extremely limited support!
  warnings.warn(
/anaconda/envs/rapids-dataloader-from-source/lib/python3.8/site-packages/merlin/schema/tags.py:148: UserWarning: Compound tags like Tags.USER_ID have been deprecated and will be removed in a future version. Please use the atomic versions of these tags, like [<Tags.USER: 'user'>, <Tags.ID: 'id'>].
  warnings.warn(
/anaconda/envs/rapids-dataloader-from-source/lib/python3.8/site-packages/merlin/schema/tags.py:148: UserWarning: Compound tags like Tags.ITEM_ID have been deprecated and will be removed in a future version. Please use the atomic versions of these tags, like [<Tags.ITEM: 'item'>, <Tags.ID: 'id'>].
  warnings.warn(
```

2nd notebook runs fine.

Then on running the imports for the third notebook, the kernel dies with:

```
Canceled future for execute_request message before replies were done
The Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click [here](https://aka.ms/vscodeJupyterKernelCrash) for more info. View Jupyter [log](command:jupyter.viewOutput) for further details.
```

**Comment by rnyak** [comment]
@masoncusack these are not errors they are just warnings. Besides, are you killing (shutting down) the kernels before you move on to the next notebook? Meaning your GPU memory is free right, when you move to the next?

**Comment by masoncusack** [comment]
Yes, but I don't want to get in the habit of ignoring warnings. A lot of them I don't understand

```
Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH:  <---Not sure whether this kind of thing is important and might break something later on or not.
```

So making sure we have full context.

I restarted and imports worked fine. Now in the second code cell (feature engineering with NVTabular) I'm getting the error

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[3], line 9
      6 SYNTHETIC_DATA = eval(os.environ.get("SYNTHETIC_DATA", "True"))
      8 if SYNTHETIC_DATA:
----> 9     train, valid = generate_data("aliccp-raw", int(NUM_ROWS), set_sizes=(0.7, 0.3))
     10     # save the datasets as parquet files
     11     train.to_ddf().to_parquet(os.path.join(DATA_FOLDER, "train"))

File /anaconda/envs/rapids-dataloader-from-source/lib/python3.8/site-packages/merlin/datasets/synthetic.py:136, in generate_data(input, num_rows, set_sizes, min_session_length, max_session_length, device)
    127         new_properties["value_count"]["max"] = max_session_length
    128     schema[col] = ColumnSchema(
    129         name=schema[col].name,
    130         tags=schema[col].tags,
   (...)
    133         is_list=True,
    134     )
--> 136 df = generate_user_item_interactions(
    137     schema, num_rows, min_session_length, max_session_length, device=device
    138 )
    140 if list(set_sizes) != [1.0]:
    141     num_rows = df.shape[0]

File /anaconda/envs/rapids-dataloader-from-source/lib/python3.8/site-packages/merlin/datasets/synthetic.py:232, in generate_user_item_interactions(schema, num_interactions, min_session_length, max_session_length, device)
    226 if user_id_cols:
    227     user_id_col = user_id_cols[0]
    228     data[user_id_col.name] = _array.clip(
    229         _array.random.lognormal(3.0, 1.0, num_interactions).astype(_array.int32),
    230         1,
    231         user_id_col.int_domain.max,
--> 232     ).astype(str(user_id_col.dtype.to_numpy.name))
    233     features = list(schema.select_by_tag(Tags.USER).remove_by_tag(Tags.USER_ID))
    234     data = generate_conditional_features(
    235         data,
    236         features,
   (...)
    240         device=device,
    241     )

AttributeError: 'numpy.dtype[int32]' object has no attribute 'to_numpy'
```

I'll look into this further tomorrow. Sorry I am just pasting errors to you but there have been so many I'm not sure what is solvable without the understanding of a maintainer and what isn't.

Is there a near-term plan to fix the installation process of this and Transformers4Rec, so we can just install a specific version and have all the sub dependencies installed and pinned? I'm concerned that installing from source exposes us to breaking changes, so probably wouldn't be possible in production.

**Comment by rnyak** [comment]
@masoncusack sorry for the inconvenience and agree on that `installing from source exposes us to breaking changes, so probably wouldn't be possible in production.`. Merlin team is working on publishing pypi versions, hopefully that'd help to the users.

**Comment by jperez999** [comment]
@masoncusack I have attached a temporary fix for your situation, the following is a snippet from a dockerfile. This is the bare minimum necessary to run the getting-started-movielens notebooks. Some of the lines you wont need because they are for installing python 3.8. The last RUN command is probably where most of your attention should fall. 

```
FROM nvcr.io/nvidia/cuda:11.8.0-devel-ubuntu22.04

RUN apt update && \
    apt install software-properties-common curl git -y && \
    add-apt-repository ppa:deadsnakes/ppa -y
RUN apt update && \
    DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt install -y \
        python3.8 \
        python3.8-dev \
        python3.8-distutils \
        # Required for Merlin
        graphviz \
        # Required for Tensorflow, to recognize GPUs
        libcudnn8 

RUN ln -s /usr/bin/python3.8 /usr/bin/python

RUN curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && \
    python3.8 get-pip.py

RUN pip install cudf-cu11 dask-cudf-cu11 --extra-index-url=https://pypi.ngc.nvidia.com && \
    # packages required for running the unit/integration tests in merlin
    pip install pytest fiddle wandb nest-asyncio testbook git+https://github.com/rapidsai/asvdb.git@main && \
    pip install tensorflow-gpu==2.9.2 scikit-learn tritonclient[all] && \ 
    pip install protobuf==3.20.3 pynvml ipython ipykernel graphviz && \
    # pip install merlin-core nvtabular merlin-models merlin-dataloader transformers4rec && \
    pip install --upgrade git+https://github.com/NVIDIA-Merlin/core.git@v0.10.0 \
                          git+https://github.com/NVIDIA-Merlin/nvtabular.git@v1.8.0 \
                          git+https://github.com/NVIDIA-Merlin/models.git@v0.11.0 \
                          git+https://github.com/NVIDIA-Merlin/dataloader.git@v0.0.4 \
                          git+https://github.com/NVIDIA-Merlin/transformers4rec.git@v0.1.16
```
If you end up having issues with tensorflow, where it says it cannot detect GPUs. Make sure to install the libcudnn package with apt. That should help with that. And if graphviz does not work, because its not in the PATH you may also want to install it via apt. With this set of instructions you should be able to make it through the third notebook of the getting started movielens. You will not be able to complete the final notebook because Tritonserver is not available in the environment. I am currently working to fix that, but that is not ready yet. 

Please let me know if this does not unblock you. Hope this helps.

**Comment by liyunrui** [comment]
I got same issues. Any workaround?

**Comment by rnyak** [comment]
> I got same issues. Any workaround?

@liyunrui could you please create another bug ticket with your issues/errors? thanks.

**Comment by liyunrui** [comment]
> ```shell
> ```shell
> pip install "merlin-models[tensorflow,nvtabular]" "protobuf>=3.20.1,<3.21" notebook
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```
I did 
Step1:conda create -y -n rapids-22.08 -c rapidsai -c nvidia -c conda-forge rapids=22.08 python=3.8 cudatoolkit=11.5
Step2: conda activate rapids-22.08
Step3: pip install "merlin-models[tensorflow,nvtabular]" "protobuf>=3.20.1,<3.21" notebook

But got error while in step3 in the below:
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
dask-cudf 22.8.1 requires cupy-cuda115<11.0.0a0,>=9.5.0, which is not installed.
cudf 22.8.1 requires cupy-cuda115<11.0.0a0,>=9.5.0, which is not installed.
cudf-kafka 22.8.1 requires cython, which is not installed.

**Comment by rnyak** [comment]
> 

are you working with Mason? if no, can you please open a new bug ticket? thanks.

**Comment by masoncusack** [comment]
> 

hi @oliverholworthy 

Looking into it but with this approach I'm currently getting the following error in the imports of notebook 3.

```
---------------------------------------------------------------------------
CUDARuntimeError                          Traceback (most recent call last)
Cell In[2], line 4
      1 import os
      2 import numpy as np
----> 4 from nvtabular.loader.tf_utils import configure_tensorflow
      6 configure_tensorflow()
      8 import nvtabular as nvt

File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/nvtabular/__init__.py:20
     18 # this is to mimic the previous API, but all of these should probably be removed
     19 import merlin.dag as graph  # noqa
---> 20 from merlin import io
     21 from merlin.core import dispatch, utils  # noqa
     22 from merlin.dag import ColumnSelector
File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/merlin/io/__init__.py:18
      1 #
      2 # Copyright (c) 2022, NVIDIA CORPORATION.
      3 #
   (...)
     16 
     17 # flake8: noqa
---> 18 from merlin.io import dataframe_iter, dataset, shuffle
     19 from merlin.io.dataframe_iter import DataFrameIter
     20 from merlin.io.dataset import Dataset

File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/merlin/io/dataset.py:33
     30 from fsspec.core import get_fs_token_paths
     31 from fsspec.utils import stringify_path
---> 33 from merlin.core.dispatch import (
     34     convert_data,
     35     hex_to_int,
     36     is_dataframe_object,
     37     is_list_dtype,
     38     list_val_dtype,
     39 )
     40 from merlin.core.utils import device_mem_size, global_dask_client, set_client_deprecated
     41 from merlin.io.csv import CSVDatasetEngine
File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/merlin/core/dispatch.py:36
     34 if HAS_GPU:
     35     try:
---> 36         import cudf
     37         import cupy as cp
     38         import dask_cudf

File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/cudf/__init__.py:5
      1 # Copyright (c) 2018-2022, NVIDIA CORPORATION.
      3 from cudf.utils.gpu_utils import validate_setup
----> 5 validate_setup()
      7 import cupy
      8 from numba import config as numba_config, cuda

File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/cudf/utils/gpu_utils.py:55, in validate_setup()
     53 except CUDARuntimeError as e:
     54     if e.status in notify_caller_errors:
---> 55         raise e
     56     # If there is no GPU detected, set `gpus_count` to -1
     57     gpus_count = -1
File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/cudf/utils/gpu_utils.py:52, in validate_setup()
     31 notify_caller_errors = {
     32     cudaError_t.cudaErrorInitializationError,
     33     cudaError_t.cudaErrorInsufficientDriver,
   (...)
     48     cudaError_t.cudaErrorApiFailureBase,
     49 }
     51 try:
---> 52     gpus_count = getDeviceCount()
     53 except CUDARuntimeError as e:
     54     if e.status in notify_caller_errors:

File /anaconda/envs/rapids-22.08/lib/python3.8/site-packages/rmm/_cuda/gpu.py:101, in getDeviceCount()
     99 status, count = cudart.cudaGetDeviceCount()
    100 if status != cudart.cudaError_t.cudaSuccess:
--> 101     raise CUDARuntimeError(status)
    102 return count

CUDARuntimeError: cudaErrorInitializationError: initialization error
```

**Comment by oliverholworthy** [comment]
@masoncusack that looks like it's failing on `import cudf`. Maybe check that you can import cudf in a separate python shell to see if it's a CUDA/RAPIDS setup issue.

**Comment by jperez999** [comment]
Hello @masoncusack if you had issues with the previous pip install version I supplied, thanks to work @oliverholworthy completed. We now have the appropriate packages available. That means the following works as well.
```
FROM nvcr.io/nvidia/cuda:11.8.0-devel-ubuntu22.04

ENV DEBIAN_FRONTEND=noninteractive

RUN apt update && \
    apt install software-properties-common curl git -y && \
    add-apt-repository ppa:deadsnakes/ppa -y
RUN apt update && \
    TZ=Etc/UTC apt install -y \
        python3.8 \
        python3.8-dev \
        python3.8-distutils \
        # Required for Merlin
        graphviz \
        # Required for Tensorflow, to recognize GPUs
        libcudnn8 

RUN ln -s /usr/bin/python3.8 /usr/bin/python

RUN curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && \
    python3.8 get-pip.py

RUN pip install cudf-cu11 dask-cudf-cu11 --extra-index-url=https://pypi.ngc.nvidia.com && \
    # packages required for running the unit/integration tests in merlin
    pip install pytest fiddle wandb nest-asyncio testbook git+https://github.com/rapidsai/asvdb.git@main && \
    pip install tensorflow-gpu==2.9.2 scikit-learn tritonclient[all] && \ 
    pip install protobuf==3.20.3 pynvml ipython ipykernel graphviz && \
    pip install merlin-core nvtabular merlin-models merlin-dataloader transformers4rec merlin-systems
```
Notice that you are now using the latest releases of the merlin libraries available on pypi. I have tested this and I have been able to get through the first 3 notebooks of the getting-started-movielens example. We cannot get through the fourth because we are missing the tritonserver install. Currently we are working on creating a tritonserver install from source.

**Comment by bschifferer** [comment]
Hello @masoncusack - does it work for you now? One QQ: In which component of Merlin are you interested? Is it Merlin Models session-based?

**Comment by bschifferer** [comment]
I tested it myself, as well. I used a clean ubuntu1804 image and installed NVIDIA Driver, etc. in the image.

- I install NVIDIA Driver 470.57.02 with CUDA 11.4 (our (support matrix)[https://nvidia-merlin.github.io/Merlin/main/support_matrix/support_matrix_merlin_tensorflow.html] shows which driver are supported)
- Note that I the 3rd code block is specific to Ubuntu1804 - that might need to be updated
- RAPIDs needs to be installed for the correct CUDA version (cudatoolkit=11.4 in the create conda command)
- I need to add links for libcublas.so and libcublasLt.so, otherwise, TensorFlow will not be able to handle RNNs and Transformers.

`sudo apt update -y && \
    sudo apt install -y build-essential && \
    sudo apt install -y --no-install-recommends software-properties-common

wget https://developer.download.nvidia.com/compute/cuda/11.4.1/local_installers/cuda_11.4.1_470.57.02_linux.run
sudo sh cuda_11.4.1_470.57.02_linux.run

wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin  && \
    sudo mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600  && \
    sudo apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/3bf863cc.pub  && \
    sudo add-apt-repository "deb https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/ /"  && \
    sudo apt-get update -y && sudo apt install -y --allow-change-held-packages git python3.8 python3-setuptools wget openssl libssl-dev zlib1g-dev libcudnn8 libcudnn8-dev

wget https://repo.anaconda.com/miniconda/Miniconda3-py38_4.10.3-Linux-x86_64.sh && bash Miniconda3-py38_4.10.3-Linux-x86_64.sh -b
~/miniconda3/bin/conda init
bash
conda create -y -n rapids -c rapidsai -c nvidia -c conda-forge rapids=22.06 python=3.8 cudatoolkit=11.4

conda activate rapids

pip install pytest testbook

git clone https://github.com/rapidsai/asvdb.git build-env && \
    pushd build-env && \
      python setup.py install && \
    popd && \
    rm -rf build-env

git clone https://github.com/NVIDIA-Merlin/core.git && cd core && git fetch origin && git checkout tags/v0.10.0 && pip install . && cd ..
git clone https://github.com/NVIDIA-Merlin/NVTabular.git && cd NVTabular && git fetch origin && git checkout tags/v1.8.0 && pip install . && cd ..
git clone https://github.com/NVIDIA-Merlin/dataloader.git && cd dataloader && git fetch origin && git checkout tags/v0.0.4 && pip install . && cd ..

pip install numpy==1.20.3
pip install tensorflow-gpu==2.10.0
git clone https://github.com/NVIDIA-Merlin/models.git && cd models && git fetch origin && git checkout tags/v0.11.0 && pip install . && cd ..
pip install fiddle wandb
pip install numpy==1.20.3
pip install protobuf==3.20.1
ln -s ~/miniconda3/envs/rapids/lib/libcublas.so.11 ~/miniconda3/envs/rapids/lib/libcublas.so.10
ln -s ~/miniconda3/envs/rapids/lib/libcublasLt.so.11 ~/miniconda3/envs/rapids/lib/libcublasLt.so.10`

**Comment by masoncusack** [comment]
Hi folks, sorry for the delay in responding.

Thanks for the input. I seem to have gotten things working as follows, based on @bschifferer's approach above:

```bash
conda create -y -n rapids -c rapidsai -c nvidia -c conda-forge rapids=22.06 python=3.8 cudatoolkit=11.4

pip install pytest testbook

git clone https://github.com/rapidsai/asvdb.git build-env
pushd build-env
python setup.py install
popd
rm -rf build-env

git clone https://github.com/NVIDIA-Merlin/core.git
cd core
git fetch origin
git checkout tags/v0.10.0 
# HEAD is now at 2fc6889c add schema parameter to the `repartition` method (#192)
pip install .
cd ..

git clone https://github.com/NVIDIA-Merlin/NVTabular.git
cd NVTabular
git fetch origin
git checkout tags/v1.8.0
pip install .
cd ..

git clone https://github.com/NVIDIA-Merlin/dataloader.git
cd dataloader
git fetch origin
git checkout tags/v0.0.4
# HEAD is now at fd5d3fc Use tf.function for list column operations (#89)
pip install .
cd ..

pip install numpy==1.20.3
pip install tensorflow-gpu==2.10.0
git clone https://github.com/NVIDIA-Merlin/models.git
cd models
git fetch origin
git checkout tags/v0.11.0
# HEAD is now at b90e9a1b Update branch name extraction for tag builds (#972)
pip install .
cd ..

pip install fiddle wandb
pip install numpy==1.20.3
pip install protobuf==3.20.1

ln -s /anaconda/envs/rapids/lib/libcublas.so.11 ~/anaconda/envs/rapids/lib/libcublas.so.10
# ln: failed to create symbolic link '/home/azureuser/anaconda/envs/rapids/lib/libcublas.so.10': No such file or directory

 ln -s /anaconda/envs/rapids/lib/libcublasLt.so.11 ~/anaconda/envs/rapids/lib/libcublasLt.so.10
# ln: failed to create symbolic link '/home/azureuser/anaconda/envs/rapids/lib/libcublasLt.so.10': No such file or directory
```

All good, but when running 4th notebook.
On import seedir as sd: 
```
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[8], line 1
----> 1 import seedir as sd
      3 sd.seedir(
      4     DATA_FOLDER,
      5     style="lines",
   (...)
      9     sort=True,
     10 )

ModuleNotFoundError: No module named 'seedir'
```

Pip install might be broken, I restarted the kernel and got the same result, so tried installing manually from the terminal

pip install seeder

Which worked.

On notebook 7 under “Training an implicit model”, I get the following error during the import:

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[12], line 1
----> 1 from merlin.models.implicit import BayesianPersonalizedRanking

File /anaconda/envs/rapids/lib/python3.8/site-packages/merlin/models/implicit/__init__.py:16
      1 #
      2 # Copyright (c) 2022, NVIDIA CORPORATION.
      3 #
   (...)
     14 # limitations under the License.
     15 #
---> 16 import implicit
     17 from implicit.evaluation import ranking_metrics_at_k
     19 from merlin.io import Dataset

ModuleNotFoundError: No module named 'implicit'
```

This is odd because the IDE detects the existence of this directory and the BayesianPersonalizedRanking class.  Happy to pursue a solution to this but I don’t think it’s going to affect our intended use case, at least for now.

As I said with the Transformers4Rec issue, this is fine for us looking into pursuing offline testing with the library and continuing to evaluate whether to use Merlin models, but this isn't an install process that we're likely to be comfortable within the context of an automated pipeline updating models in production. So hopefully the ability for us to simply pip install merlin-models or similar is on the roadmap.

Huge thanks for all the input.

**Comment by viswa-nvidia** [comment]
@rnyak , can this bug be closed

**Comment by rnyak** [comment]
@masoncusack we are glad that you can install the libs. And thanks to providing your steps for those who might go through the same issues. Just to clarify, I assume you did not install `NV drivers` and and `nvidia docker toolkit` right? Your Azure instance come with these installed, am I correct?

Can you please create a separate ticket for the `ModuleNotFoundError: No module named 'implicit'`. I am closing this ticket.

**Comment by dking21st** [comment]
@jperez999 
Hi, I initially did my merlin modeling on Google Colab, and now trying to deploy the model training & upload job on airflow KubernetePodOperator with docker image.
I've been struggling since 3 weeks ago - can you help me figuring out what base image to use if Iwould like to use following merlin versions? 

!pip install -U git+https://github.com/NVIDIA-Merlin/models.git@release-23.04
!pip install -U git+https://github.com/NVIDIA-Merlin/nvtabular.git@release-23.04
!pip install -U git+https://github.com/NVIDIA-Merlin/core.git@release-23.04
!pip install -U git+https://github.com/NVIDIA-Merlin/dataloader.git@release-23.04
!pip install merlin-systems==23.04
!pip install tensorflow==2.12.0
